File: dbutil/makeTask.rb
require 'net/http'
require 'uri'
require 'json'
require 'securerandom'
uuid = SecureRandom.uuid

# Include config variables - BB 
require './Configuration' 
include Configuration 
 
DNS = Configuration::DNS 
DB_PORT = Configuration::DB_PORT 
DB_ADMIN_USER = Configuration::DB_ADMIN_USER 
DB_ADMIN_PASS = Configuration::DB_ADMIN_PASS 
IMAGES_DB = Configuration::IMAGES_DB
ADMIN_PARTY = Configuration::ADMIN_PARTY


# Include config variables - BB


# get the number of documents as a command line arg
ARGV.each { |arg| puts "Argument: #{arg}" }
user = ARGV[0]
i_list = ARGV[1]
task_type=ARGV[2]
task_order=ARGV[3] #.to_int

description = "Default description that should be changed."
description=ARGV[4] if ARGV[4]

if (ARGV.size <3) then
  puts "Usage: ruby : #{$PROGRAM_NAME} <user> <image-list name> <image-list-type> <task-order> [<description>]"
  puts "Where:\n"
  puts "1 <user> is the user assigned to the task.\n"
  puts "2 <image-list name> is an Image Compare/Classify List created before this step.\n"
  puts "3 <image-list-type> is compare, OCTcompare, classify, classify_nine or quandrant.\n"
  puts "4 <task-order> is the order in which to complete this task relatve to other tasks.\n"
  puts "5 <description> Description that will be above images.\n"
  exit
end

#if task order is provided, use it otherwise task order is set to arbitrarily high number so that it shows up last
if (ARGV.size > 3)
  task_order=task_order.strip.to_i
  puts task_order
  puts task_order.class
else
  task_order=100
end


obj = { type:"task",
        task_type:task_type,
        task_order:task_order,
        user:user,
        timeAdded:Time.now(),
  #    image_compare_list:image_compare_list,
  #      image_classify_list:image_classify_list,
        current_idx:0,
        completed:false}


obj["list_name"]=i_list


# Delete if we don't need anymore
# if task_type=="compare"
#   obj["image_compare_list"]=i_list
# elsif task_type=="classify" || task_type=="classify_nine" || task_type=="quadrant"
#   obj["image_classify_list"]=i_list
# elsif task_type=="OCTcompare"
#   obj["OCTimage_compare_list"]=i_list
# else
#   puts "refer to help2"
#   exit
# end


# BB - not sure if this is important - #

# need to change icl in db to reflect
#puts obj.inspect
#puts obj.to_json

# BB - not sure if this is important - #


# put the results in the database
docname = uuid

url = "http://#{DNS}:#{DB_PORT}/" + IMAGES_DB + "/" + docname
puts url

####
uri = URI.parse(url)

http = Net::HTTP.new(uri.host, uri.port)
request = Net::HTTP::Put.new(uri.request_uri)
if ADMIN_PARTY != true
  request.basic_auth(DB_ADMIN_USER, DB_ADMIN_PASS)
end
response = http.request(request, obj.to_json)
response = JSON.parse(response.body)
puts response
#####

File: dbutil/makeTask.py
import os, sys, requests, json, couchdb, uuid, pdb, pprint as pp
from datetime import datetime, timezone, timedelta
from dotenv import load_dotenv

load_dotenv("../flask_server/.env",verbose=True)
 
DB_ADMIN_USER = os.getenv("DB_ADMIN_USER")
DB_ADMIN_PASS = os.getenv("DB_ADMIN_PASS")
DNS = os.getenv("DNS")
IMAGES_DB = os.getenv("IMAGES_DB")
DB_PORT = os.getenv("DB_PORT")
HTTP_PORT = os.getenv("HTTP_PORT")
ADMIN_PARTY = True if os.getenv("ADMIN_PARTY") == 'True' else False

# https://couchdb-python.readthedocs.io/en/latest/getting-started.html
if ADMIN_PARTY:
    couch = couchdb.Server(f'http://{DNS}:{DB_PORT}')
else:
    couch = couchdb.Server(f'http://{DB_ADMIN_USER}:{DB_ADMIN_PASS}@{DNS}:{DB_PORT}')

# couch package ex for later
    # db = couch[IMAGES_DB]
    # imageIDs = [int(row['id']) for row in db.view('_design/basic_views/_view/imageSet2ImageId')]
    # imageIDs.sort()
    # imageIDs = [str(i) for i in imageIDs]

# def getURL(uuid: str) -> str:
#     url = f"http://{DNS}:{DB_PORT}/{IMAGES_DB}"
#     view = f'/{uuid}'
#     URL = url + view
#     return URL

def makeTask(user: str,imageListName: str, imageListType: str, taskOrder: int) -> None:
    t = datetime.now() - timedelta(hours=4)
    obj = {"type":"task",
           "task_type":imageListType, 
           "list_name": imageListName,
           "task_order": taskOrder,
           "user":user,
           "time_added":t.strftime('%Y-%m-%d %H:%M:%S'),
           "current_idx":0,
           "completed":False}
    db = couch[IMAGES_DB]
    # pdb.set_trace()
    doc_id, doc_rev = db.save(obj)
    print(pp.pprint(f"created object {obj}"))
    
def main(user: str,imageListName: str, imageListType: str, taskOrder: int):
    # uid = uuid.uuid1()
    # url = getURL(str(uid))
    # pdb.set_trace()
    makeTask(user, imageListName, imageListType, taskOrder)



if __name__ == "__main__":
    try:
        main(sys.argv[1], sys.argv[2],sys.argv[3],sys.argv[4])
    except IndexError as err:
        print(f"""
        Error: {err}, and probably means you 
        didn't provide <user> or <imageListName> <imageListType> <taskOrder>
        """)


